torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1150,11 +1158,93 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   // threads dying due to aborted communicator and raising a SIGABRT
   std::string abortReason = c10::str("Process Group destroyed on rank ", rank_);
   abort(abortReason);
+
+  // We need to wait for abort to finish before we can safely shut down
+  // heartbeat monitoring thread.
+  terminateHeartbeatMonitorThread_.store(true);
```

Comment:
hmm. do we have good enough control over heartbeatMonitor so that it waits long enough after `terminateProcessGroup` is set true, before abort() finishes?

1. What happens inside watchdog thread when we call terminate? 
         Answer: we continue processing a 'for loop' over work objects, and possibly, issue a desync report
3. is the time to terminate + abort expected to be the same as the normal loop period? 
         Answer: not sure how long abort is, but, for-loop over work can be arbitrarily long, so we probably have enough padding in our usual heartbeat period if its safe for for-loop over work, to also account for the abort call

Reply:
For No.1, we make the heartbeat monitoring wait for long enough so that we keep enough buffer. (We probably need to bump the heartbeat inside desync report generation?)

For No.2, for now we don't really directly terminate inside watchdog, we throw exceptions after generating desync report. (Unless users want to call abort intentionally, like set `NCCL_ASYNC_ERROR_HANDLING` to be 1.)

Reply:
OK, with two flags added, I am confident that we should be able to capture most corner cases.. And fingers crossed.